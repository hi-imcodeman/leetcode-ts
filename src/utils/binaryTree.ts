export type NumberOrNull = number | null
// Definition for a binary tree node.
export class TreeNode {
    val: NumberOrNull
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number | null, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? null : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

export function insertLevelOrder(data:NumberOrNull[],i=0):TreeNode|null{
    let root = null
    if(i<data.length){
        root = new TreeNode(data[i])
        root.left = insertLevelOrder(data,2*i+1)
        root.right = insertLevelOrder(data,2*i+2)
    }
    return root
}

export function inOrder(root:TreeNode|null,arr:NumberOrNull[]=[]){
    if(root !== null){
        inOrder(root.left,arr)
        arr.push(root.val)
        inOrder(root.right,arr)
    }
    return arr
}

export function preOrder(root:TreeNode|null,arr:NumberOrNull[]=[]){
    if(root !== null){
        arr.push(root.val)
        inOrder(root.left,arr)
        inOrder(root.right,arr)
    }
    return arr
}

export function postOrder(root:TreeNode|null,arr:NumberOrNull[]=[]){
    if(root !== null){
        inOrder(root.left,arr)
        inOrder(root.right,arr)
        arr.push(root.val)
    }
    return arr
}

export function levelOrder(root:TreeNode|null,arr:Array<NumberOrNull[]>=[],level=0){
    if(root){
        if(!arr[level]){
            arr[level]=[]
        }
        arr[level].push(root.val)
        levelOrder(root.left,arr,level+1)
        levelOrder(root.right,arr,level+1)
    }
    return arr
}