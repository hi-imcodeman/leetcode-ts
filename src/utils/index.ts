export * from './linkedList'
export * from './binaryTree'
