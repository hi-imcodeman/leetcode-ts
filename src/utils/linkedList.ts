
//Definition for singly-linked list.
export class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

export const arrayToLinkedList = (nums:number[]):ListNode | null => {
    if(!nums.length)
        return null
    const listNode:ListNode = new ListNode(nums[0])
        let currNode = listNode
        let i=1
        while(i<nums.length){
            const node = new ListNode(nums[i])
            currNode.next = node
            currNode = currNode.next
            i++
        }
        return listNode
}

export const linkedListToArray = (listNode:ListNode|null):number[] => {
    if(!listNode)
        return []
    const data:number[]=[]
    let currNode = listNode
    data.push(currNode.val)
    while(currNode.next){
        currNode = currNode.next
        data.push(currNode.val)
    }
    return data
}

