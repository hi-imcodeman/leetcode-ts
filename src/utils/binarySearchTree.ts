import { TreeNode } from './binaryTree'

export function insertLevelOrderBST(arr: number[]|string[]): TreeNode<number|string> | null {
    let root:TreeNode<number|string>|null = null
    const insert = (value:number|string,node:TreeNode<number|string>|null)=>{
        if(node === null)
            return new TreeNode(value)
        if(value < node.val){
            node.left = insert(value,node.left)
        }else if(value >= node.val){
            node.right = insert(value,node.right)
        }
        return node
    }
    arr.forEach((item:number|string)=>{
        root = insert(item,root)
    })
    return root
}