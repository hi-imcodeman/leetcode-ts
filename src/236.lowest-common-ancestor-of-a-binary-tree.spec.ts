import { lowestCommonAncestor } from './236.lowest-common-ancestor-of-a-binary-tree'
import { insertLevelOrder } from './utils'

describe('236. Lowest common Ancestor of a binary tree', () => {
    test('Case 1', () => {
        const root = insertLevelOrder([3,5,1,6,2,0,8,null,null,7,4])
        const p = insertLevelOrder([5,6,2,null,null,7,4])
        const q = insertLevelOrder([1,0,8])
        const result = lowestCommonAncestor(root,p,q)
        expect(result).toMatchSnapshot()
    })
    test('Case 2', () => {
        const root = insertLevelOrder([3,5,1,6,2,0,8,null,null,7,4])
        const p = insertLevelOrder([5,6,2,null,null,7,4])
        const q = insertLevelOrder([5])
        const result = lowestCommonAncestor(root,p,q)
        expect(result).toEqual(p)
    })
    test('Case 3', () => {
        const root = insertLevelOrder([1,2])
        const p = insertLevelOrder([1,2])
        const q = insertLevelOrder([2])
        const result = lowestCommonAncestor(root,p,q)
        expect(result).toEqual(p)
    })
})