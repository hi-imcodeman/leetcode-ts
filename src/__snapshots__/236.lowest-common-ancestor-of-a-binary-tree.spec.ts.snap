// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`236. Lowest common Ancestor of a binary tree Case 1 1`] = `
TreeNode {
  "left": TreeNode {
    "left": TreeNode {
      "left": null,
      "right": null,
      "val": 6,
    },
    "right": TreeNode {
      "left": TreeNode {
        "left": null,
        "right": null,
        "val": 7,
      },
      "right": TreeNode {
        "left": null,
        "right": null,
        "val": 4,
      },
      "val": 2,
    },
    "val": 5,
  },
  "right": TreeNode {
    "left": TreeNode {
      "left": null,
      "right": null,
      "val": 0,
    },
    "right": TreeNode {
      "left": null,
      "right": null,
      "val": 8,
    },
    "val": 1,
  },
  "val": 3,
}
`;
